cmd/mel-replay/delayed_message_db_test.go

```diff
@@ -0,0 +1,192 @@
+		for i, expected := range expectedData {
+			msg, err := db.ReadDelayedMessage(ctx, state, uint64(i)) // #nosec G115
+			require.NoError(t, err)
+			require.Equal(t, expected, msg.Message.L2msg)
+		}
+	})
+	t.Run("Merkle tree with 3 levels can fetch specific delayed messages", func(t *testing.T) {
+		// We have a Merkle tree for delayed messages that looks like this:
+		//
+		//     hash(hash(A++B)++hash(C++D))
+		//       /                \
+		//   hash(A++B)        hash(C++D)
+		//   /       \          /       \
+		//  A         B        C         D
+		//
+		// We should be able to fetch A, B, C, or D.
+		messages := []*mel.DelayedInboxMessage{
+			buildDelayedMessage(t, 1, []byte("a")),
+			buildDelayedMessage(t, 2, []byte("b")),
+			buildDelayedMessage(t, 3, []byte("c")),
+			buildDelayedMessage(t, 4, []byte("d")),
+		}
+
+		preimages, root := buildMerkleTree(t, messages)
+
+		resolver := &mockPreimageResolver{preimages: preimages}
+		db := &delayedMessageDatabase{preimageResolver: resolver}
+		state := &mel.State{
+			DelayedMessagedSeen:     4,
+			DelayedMessagesSeenRoot: root,
+		}
+
+		// Test each message
+		expectedData := [][]byte{[]byte("a"), []byte("b"), []byte("c"), []byte("d")}
+		for i, expected := range expectedData {
+			msg, err := db.ReadDelayedMessage(ctx, state, uint64(i)) // #nosec G115
+			require.NoError(t, err)
+			require.Equal(t, expected, msg.Message.L2msg)
+		}
+	})
+}
+
+func TestNextPowerOfTwo(t *testing.T) {
+	testCases := []struct {
+		input    uint64
+		expected uint64
+	}{
+		{0, 1},
+		{1, 1},
+		{2, 2},
+		{3, 4},
+		{4, 4},
+		{5, 8},
+		{8, 8},
+		{9, 16},
+		{16, 16},
+		{17, 32},
+	}
+
+	for _, tc := range testCases {
+		result := nextPowerOfTwo(tc.input)
+		if result != tc.expected {
+			t.Errorf("nextPowerOfTwo(%d) = %d, expected %d", tc.input, result, tc.expected)
+		}
+	}
+}
+
+func buildDelayedMessage(
+	_ *testing.T,
+	blockNumber uint64,
+	msgData []byte,
+) *mel.DelayedInboxMessage {
+	msg := &mel.DelayedInboxMessage{
+		ParentChainBlockNumber: blockNumber,
+		Message: &arbostypes.L1IncomingMessage{
+			Header: &arbostypes.L1IncomingMessageHeader{
+				Kind: arbostypes.L1MessageType_L2Message,
+			},
+			L2msg: msgData,
+		},
+	}
+	return msg
+}
+
+func buildMerkleTree(t *testing.T, messages []*mel.DelayedInboxMessage) (map[common.Hash][]byte, common.Hash) {
+	preimages := make(map[common.Hash][]byte)
+	// Encode messages and get leaf hashes.
+	leafHashes := make([]common.Hash, len(messages))
+	for i, msg := range messages {
+		encoded, err := rlp.EncodeToBytes(msg)
+		require.NoError(t, err)
+		hash := crypto.Keccak256Hash(encoded)
+		preimages[hash] = encoded
+		leafHashes[i] = hash
+	}
+	// Build tree bottom-up
+	currentLevel := leafHashes
+	for len(currentLevel) > 1 {
+		nextLevel := make([]common.Hash, 0, len(currentLevel)/2)
+
+		for i := 0; i < len(currentLevel); i += 2 {
+			left := currentLevel[i]
+			right := currentLevel[i+1] // Assumes even number of leaves
```

Comment:
that's a good point.
Let's either add or change last test to have odd number of leaves :)